File: javascript/02_contract_call.js
#!/usr/bin/env node
/**
 * Example 2 — QVM contract call (qweb3.js)
 *
 * Read from and write to a QRC20 token on the Quantova Virtual Machine. The QVM
 * uses the standard Solidity ABI (keccak-256 selectors), so this is identical to
 * EVM tooling — only the transaction signature is post-quantum.
 *
 * Run:
 *   export QUANTOVA_RPC=http://127.0.0.1:9933
 *   node 02_contract_call.js
 *
 * In a full app with qweb3.js installed:
 *   const { QWeb3 } = require('qweb3.js');
 *   const q = new QWeb3(process.env.QUANTOVA_RPC);
 *   const token = q.contract(QRC20_ABI, TOKEN);
 *   const bal = await token.call('balanceOf', [HOLDER]);         // decoded automatically
 *   const data = token.encode('transfer', [HOLDER, 25n * 10n**18n]);
 *   AbiCodec.functionSelector('transfer(address,uint256)')        // '0xa9059cbb'
 */
'use strict';
const { rpc, num, banner } = require('./_common');

const TOKEN = '0xC0ffee254729296a45a3885639AC7E10F9d54979';
const HOLDER = '0x1111111111111111111111111111111111111111';

// keccak-256 selectors/topics (qweb3.js computes these via AbiCodec; shown as
// constants here so the example runs without a keccak dependency).
const SEL_BALANCE_OF = '0x70a08231'; // balanceOf(address)
const SEL_TRANSFER = '0xa9059cbb';   // transfer(address,uint256)
const TOPIC_TRANSFER = '0xddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3ef';

function pad32(hexNo0x) { return hexNo0x.replace(/^0x/, '').padStart(64, '0'); }

async function main() {
  banner('Quantova example: QVM contract call (QRC20) (qweb3.js)');
  console.log(`Connected. Block: ${num(await rpc('q_blockNumber'))}`);
  console.log(`Token contract: ${TOKEN}`);
  console.log(`selector transfer(address,uint256): ${SEL_TRANSFER}`);
  console.log(`topic    Transfer(address,address,uint256): ${TOPIC_TRANSFER.slice(0, 18)}…`);

  // read: balanceOf(HOLDER) via q_call
  const callData = SEL_BALANCE_OF + pad32(HOLDER);
  const ret = await rpc('q_call', [{ to: TOKEN, data: callData }]);
  const bal = BigInt(ret);
  console.log(`\nbalanceOf(${HOLDER.slice(0, 10)}…) = ${bal}  (= ${bal / 10n ** 18n} tokens)`);

  // write: transfer 25 tokens -> calldata
  const amount = 25n * 10n ** 18n;
  const data = SEL_TRANSFER + pad32(HOLDER) + pad32(amount.toString(16));
  console.log(`\nWrite: transfer 25 tokens`);
  console.log(`calldata: 0x${data.slice(0, 38)}…`);
  console.log('In qweb3.js the wallet post-quantum-signs this calldata, then broadcasts.');

  const txHash = await rpc('q_sendRawTransaction', ['0x' + data]);
  console.log(`Broadcast! tx hash: ${txHash}`);
  const rcpt = await rpc('q_getTransactionReceipt', [txHash]);
  console.log(`status: ${rcpt.status}  gasUsed: ${num(rcpt.gasUsed)}`);
}

main().catch((e) => { console.error('error:', e.message); process.exit(1); });

File: python/02_contract_call.py
#!/usr/bin/env python3
"""Example 2 — QVM contract call (qweb3.py)

Read from and write to a QRC20 token on the Quantova Virtual Machine. Shows the
standard Solidity ABI (keccak-256 selectors), a read via q_call (decoded
automatically), and building + post-quantum-signing a write (transfer).

Run:
    export QUANTOVA_RPC=http://127.0.0.1:9933
    python 02_contract_call.py
"""

from _common import banner, connect, num, use_demo_backend

TOKEN = "0xC0ffee254729296a45a3885639AC7E10F9d54979"  # example QRC20 contract (H160)
HOLDER = "0x1111111111111111111111111111111111111111"

QRC20_ABI = [
    {"type": "function", "name": "name", "stateMutability": "view", "inputs": [], "outputs": [{"type": "string"}]},
    {"type": "function", "name": "balanceOf", "stateMutability": "view",
     "inputs": [{"name": "owner", "type": "address"}], "outputs": [{"type": "uint256"}]},
    {"type": "function", "name": "transfer", "stateMutability": "nonpayable",
     "inputs": [{"name": "to", "type": "address"}, {"name": "value", "type": "uint256"}],
     "outputs": [{"type": "bool"}]},
    {"type": "event", "name": "Transfer", "inputs": [
        {"name": "from", "type": "address", "indexed": True},
        {"name": "to", "type": "address", "indexed": True},
        {"name": "value", "type": "uint256", "indexed": False}]},
]


def main():
    banner("Quantova example: QVM contract call (QRC20)")
    use_demo_backend()
    q = connect()
    print(f"Connected. Block: {num(q.rpc.block_number())}")

    token = q.contract(QRC20_ABI, TOKEN)
    print(f"Token contract: {TOKEN}")

    # --- ABI selectors (standard keccak-256, EVM-compatible) ---
    from qweb3 import abi
    print(f"selector transfer(address,uint256): {abi.function_selector('transfer(address,uint256)')}")
    print(f"topic    Transfer(address,address,uint256): {abi.event_topic('Transfer(address,address,uint256)')[:18]}…")

    # --- read (q_call, decoded automatically) ---
    bal = token.call("balanceOf", [HOLDER])
    print(f"\nbalanceOf({HOLDER[:10]}…) = {bal}  (= {bal / 10**18} tokens)")

    # --- build + post-quantum-sign a write (transfer 25 tokens) ---
    sender = q.wallet.create("falcon")
    calldata = token.encode("transfer", [HOLDER, 25 * 10**18])
    print(f"\nWrite: transfer 25 tokens, from {sender.address}")
    print(f"calldata: {calldata[:42]}…")
    sig = q.wallet.sign_transaction(calldata, sender.address)
    print(f"Post-quantum signature: {sig[:34]}… ({len(sig)//2 - 1} bytes)")

    try:
        tx_hash = q.rpc.call("q_sendRawTransaction", [sig])
        print(f"Broadcast! tx hash: {tx_hash}")
        rcpt = q.rpc.get_transaction_receipt(tx_hash)
        print(f"status: {rcpt['status']}  gasUsed: {num(rcpt['gasUsed'])}")
    except Exception as e:
        print(f"(broadcast step: {e})")


if __name__ == "__main__":
    main()
